### ISSUE TITLE
Enable reuse of the constituent primitive key shares of hybrid KEM

### ISSUE THREAD

**Issue by baentsch** [openssl/openssl#25889]
As a follow-up feature to be considered after #25884 has been implemented:

[draft-ietf-tls-hybrid-design-11](https://datatracker.ietf.org/doc/html/draft-ietf-tls-hybrid-design-11) explicitly allows reuse of the constituent primitive key shares of hybrid KEM in the ClientHello.

I.e. sending the exact same encapsulation key for ML-KEM-768 when sending key shares for both X25519MLKEM768 and SecP256r1MLKEM768 is explicitly allowed. Same for the X25519 or P-256 key share.

This behaviour is actually already implemented in NSS.

_Originally posted by @tomato42 in https://github.com/openssl/openssl/issues/25884#issuecomment-2459767477_

**Comment by t8m** [comment]
Basically it requires special handling in libssl and exporting/importing the individual part keys of the hybrid composite from/to the provider.

I.e., in scenario with 4 key shares (X25519MLKEM768, SecP256r1MLKEM768, X25519, P-256) 1. libssl generates a key for X25519, creates a key share and exports the private key, 2. same for P-256, 3. libssl generates a key for MLKEM768, doesn't create a key share but exports the private key, 4. libssl imports the X25519 and MLKEM768 private key components into the hybrid X25519MLKEM768 EVP_PKEY and creates a key share 5. libssl imports the P-256 and MLKEM768 private key components into the hybrid SecP256r1MLKEM768 EVP_PKEY and creates a key share. There is not much to be changed in suggested APIs for the generic hybrid KEM support. The import of the private key components can be done with a single EVP_PKEY_fromdata() if the exported OSSL_PARAM arrays are merged by libssl (we have helper functions for that). I do not think we want to add another provider key management import function allowing import from multiple OSSL_PARAM arrays.

As you can see all the special handling is done on the libssl side and there is no need for any extra provider functions. Of course in case the provider does not allow exporting the private key this optimization will not work and such failure needs to be handled gracefully by libssl.

In my opinion this optimization is really low priority and not worth it. We still do not even support sending multiple key shares from the client at all which is a much bigger deficiency.

**Comment by tomato42** [comment]
> In my opinion this optimization is really low priority and not worth it. We still do not even support sending multiple key shares from the client at all which is a much bigger deficiency.

well, this has changed now... 🙂 with not only client sending multiple, but also allowing the user to set up which ones need to be sent

**Comment by t8m** [comment]
Yes, true, and it would be nice if we could get this kind of optimization added however it is still very much non-trivial effort (esp. API-wise).
